XenAdmin/TabPages/ManageUpdatesPage.cs

```diff
@@ -215,6 +220,243 @@ private bool FilterIsOn
+                                    // paint background color on icon
+                                    using (var brush = new SolidBrush(DataGridView.DefaultCellStyle.SelectionBackColor))
+                                        graphics.FillRectangle(
+                                            brush, cellBounds.X,
+                                            cellBounds.Y, hostIcon.Width, cellBounds.Height);
+                                }
+                                else       // row is not selected
+                                {
+                                    using (var brush = new SolidBrush(DataGridView.DefaultCellStyle.BackColor))
+                                        graphics.FillRectangle(
+                                            brush, cellBounds.X,
+                                            cellBounds.Y, hostIcon.Width, cellBounds.Height);
+                                }
+
+                                // paint icon
+                                graphics.DrawImage(hostIcon, cellBounds.X, cellBounds.Y + 3, hostIcon.Width,
+                                                       hostIcon.Height);
+                            }
+                        }
+                    }
+                }
+            }
+
+            private DataGridViewCell _poolIconCell;
+            private DataGridViewTextBoxCell _versionCell;
+            private DataGridViewImageCell _patchingStatusCell;
+            private DataGridViewTextBoxCell _requiredUpdateCell;
+            private DataGridViewTextBoxCell _installedUpdateCell;
+
+            public UpdatePageDataGridViewRow(Pool pool)
+                : base(pool)
+            {
+                SetupCells();
+            }
+
+            public UpdatePageDataGridViewRow(Host host, bool hasPool)
+                : base(host, hasPool)
+            {
+                SetupCells();
+            }
+
+            public bool IsPoolOrStandaloneHost
+            {
+                get { return IsAPoolRow || (IsAHostRow && !HasPool); }
+            }
+
+            private void SetupCells()
+            {
+                if (IsPoolOrStandaloneHost)
+                    _poolIconCell = new DataGridViewImageCell();
+                else
+                    _poolIconCell = new DataGridViewTextBoxCell();
+
+                _nameCell = new UpdatePageDataGridViewNameCell();
+                _versionCell = new DataGridViewTextBoxCell();
+                _patchingStatusCell = new DataGridViewImageCell();
+                _requiredUpdateCell = new DataGridViewTextBoxCell();
+                _installedUpdateCell = new DataGridViewTextBoxCell();
+
+                Cells.AddRange(new[] { _poolIconCell, _nameCell, _versionCell, _patchingStatusCell, _requiredUpdateCell, _installedUpdateCell });
+
+                this.UpdateDetails();
+            }
+
+            // fill data into row
+            private void UpdateDetails()
+            {
+                if (Tag is Pool)

```

Comment:
To avoid a double cast, it's better to check line in line 238. Same with the host below.